From 49e472c091859a921aade7d327a442b7f648a602 Mon Sep 17 00:00:00 2001
From: Alex Drozdoff <anomalocaribd@gmail.com>
Date: Thu, 26 Mar 2026 09:40:02 +1300
Subject: [PATCH 07/10] enlarge thumbs for viewing

---
 src/thumb-standard.cc | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/src/thumb-standard.cc b/src/thumb-standard.cc
index 91179487..47a8e534 100644
--- a/src/thumb-standard.cc
+++ b/src/thumb-standard.cc
@@ -69,7 +69,7 @@ struct ExifData;
 
 enum {
 	THUMB_SIZE_NORMAL = 128,
-	THUMB_SIZE_LARGE =  256
+	THUMB_SIZE_LARGE =  128
 };
 
 #define THUMB_MARKER_URI    "tEXt::Thumb::URI"
@@ -595,12 +595,6 @@ static GdkPixbuf *thumb_loader_std_finish(ThumbLoaderStd *tl, GdkPixbuf *pixbuf,
 			}
 		}
 
-	if (sw <= tl->requested_width && sh <= tl->requested_height)
-		{
-		result = pixbuf;
-		g_object_ref(result);
-		}
-	else
 		{
 		gint thumb_w;
 		gint thumb_h;
-- 
2.53.0

